ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/DB/HdWallet/Delete.hs

```diff
@@ -6,19 +8,46 @@ module Ariadne.Wallet.Cardano.Kernel.DB.HdWallet.Delete (
 
 import Universum
 
-import Control.Lens (at, (.=))
-
 import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
 import Ariadne.Wallet.Cardano.Kernel.DB.Util.AcidState
+import Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet
+
+data Selected a where
+    RootId :: HdRootId -> Selected HdRootId 
+    AccountId :: HdAccountId -> Selected HdAccountId
+
+type family UnknownId a where 
+    UnknownId HdRootId = UnknownHdRoot 
+    UnknownId HdAccountId = UnknownHdAccount 
 
 {-------------------------------------------------------------------------------
   DELETE
 -------------------------------------------------------------------------------}
 
--- | Delete a wallet
-deleteHdRoot :: HdRootId -> Update' HdWallets e ()
-deleteHdRoot rootId = zoom hdWalletsRoots $ at rootId .= Nothing
+-- | Delete a wallet with the whole subtree (addresses and accounts)
+deleteHdRoot :: HdRootId -> Update' HdWallets UnknownHdRoot ()
+deleteHdRoot rootId = do 
+  checkExistence $ RootId rootId 
+  zoom hdWalletsAddresses $ modify (deleteIxAll rootId) 
+  zoom hdWalletsAccounts  $ modify (deleteIxAll rootId) 
+  zoom hdWalletsRoots     $ modify (deleteIxAll rootId) 
+
+-- | Delete an account with it's addresses
+deleteHdAccount :: HdAccountId -> Update' HdWallets UnknownHdAccount ()
+deleteHdAccount accId = do
+  checkExistence $ AccountId accId
+  zoom hdWalletsAddresses $ modify (deleteIxAll accId)
+  zoom hdWalletsAccounts  $ modify (deleteIxAll accId) 
+
 
--- | Delete an account
-deleteHdAccount :: HdAccountId -> Update' HdWallets UnknownHdRoot ()
-deleteHdAccount accId = zoom hdWalletsAccounts $ at accId .= Nothing
+checkExistence :: Selected a -> Update' HdWallets (UnknownId a) ()
+checkExistence (RootId rootId) = 
```

Comment:
It seems to me that this function is unnecessary complication. Its body is defined for each case separately, so you could instead define two functions and their definitions would take the same amount of code. The only thing it buys is that you don't need to type its signature twice, but instead you need to define an extra data type and a type family.